Write `bluetooth-agent` pid on process start.

Qapf opened the pull request:
Write `bluetooth-agent` pid on process start.

I've experienced the `Pair a Bluetooth device manually` failing to list
any devices.

Upon inspection, `/run/bluetooth-agent.pid` is not getting the PID of
the agent written into it, making `do_devlist` in `rocknix-bluetooth`
fail.

With this change, the pid will be written out by
`rocknix-bluetooth-agent` on start.

I have also cleaned up the call to pgrep in `rocknix-bluetooth` to
ensure that the PID is being set only by the agent.

Tested on a Retroid Pocket 5 by writing bind mounting the two modified
files and starting / stopping bluetooth.

loki666 commented:
Thanks for investigating this issue.
But this will adds a dependcy on systemd to be responsible to creates the .pid file? and systemd doesn't care about that file...

would it be possible instead, to have bluetooth-agent save its pid in the file ?

Qapf commented:
> Thanks for investigating this issue. But this will adds a dependcy on systemd to be responsible to creates the .pid file? and systemd doesn't care about that file...
> 
> would it be possible instead, to have bluetooth-agent save its pid in the file ?

Sure. Might be a week before I get a free day, but entirely doable.

loki666 commented:
Thanks. no rush!

Also for a general understanding of the issue, I'm wondering how Bluetooth scan is working at all without this PID file?

It may be bugged, but usually I'm able to list my bluetooth devices (sometimes after multiple tries).

spycat88 commented:
PID is normally set [here](https://github.com/ROCKNIX/distribution/blob/ef607097ebcc1a142b415e9a9c932cdd402451d4/packages/rocknix/sources/scripts/rocknix-bluetooth#L92)

The pid is set correctly but I have observed that if a scan is initiated too quickly after enabling Bluetooth then the agent is killed. It obviously needs a rework.

You can see this by doing `systemctl status bluetooth-agent` after enabling Bluetooth and then again if you initate a scan too quickly you will see the process is killed.

porschemad911 commented:
From what I can see `/run/bluetooth-agent.pid` is created via `rocknix-bluetooth enable` here: https://github.com/ROCKNIX/distribution/blob/ef607097ebcc1a142b415e9a9c932cdd402451d4/packages/rocknix/sources/scripts/rocknix-bluetooth#L169

This is triggered on boot from the bluetooth autostart script if the `controllers.bluetooth.enabled` system setting is set to 1: https://github.com/ROCKNIX/distribution/blob/ef607097ebcc1a142b415e9a9c932cdd402451d4/packages/network/bluez/autostart/009-bluetooth#L9

It is also triggered when bluetooth is enabled from the ES menu: https://github.com/ROCKNIX/emulationstation-next/blob/421832d38fb4375145df3783ea36fad3c6e49637/es-app/src/ApiSystem.cpp#L468

@Qapf In which scenario is `/run/bluetooth-agent.pid` not being created? After a boot with bluetooth enabled? Or after a boot with bluetooth disabled, and then toggled to enabled in the ES menu?

Qapf commented:
> @Qapf In which scenario is /run/bluetooth-agent.pid not being created? After a boot with bluetooth enabled? Or after a boot with bluetooth disabled, and then toggled to enabled in the ES menu?

After a boot with bluetooth enabled. 

As well, after experiencing that failure, I toggled Bluetooth to `off`, waited 30s, toggled bluetooth back to `on`, waited 30s. I still did not observe the PID being written. 

I'm happy to implement whatever fix is agreed upon and clean up other places where the PID is being written if that's the direction we wanna go.

spycat88 commented:
> As well, after experiencing that failure, I toggled Bluetooth to `off`, waited 30s, toggled bluetooth back to `on`, waited 30s. I still did not observe the PID being written.

Strange because it is there for me, can you make sure your bluetooth adapter has a valid mac?

Qapf commented:
> > As well, after experiencing that failure, I toggled Bluetooth to `off`, waited 30s, toggled bluetooth back to `on`, waited 30s. I still did not observe the PID being written.
> 
> Strange because it is there for me, can you make sure your bluetooth adapter has a valid mac?

I believe it does

```
SD865:~ # hciconfig -a
hci0:	Type: Primary  Bus: UART
	BD Address: 00:03:7F:39:00:24  ACL MTU: 1024:7  SCO MTU: 240:4
	UP RUNNING PSCAN
```

I'm going to sit down this evening and work on making `rocknix-bluetooth-agent` write out its own PID instead of (ab)using systemd to do the same thing. I'll also clean up any other place I find in the codebase where `/run/bluetooth-agent.pid` gets written to.

Qapf committed:
Write `bluetooth-agent` pid on process start.

I've experienced the `Pair a Bluetooth device manually` failing to list
any devices.

Upon inspection, `/run/bluetooth-agent.pid` is not getting the PID of
the agent written into it, making `do_devlist` in `rocknix-bluetooth`
fail.

With this change, the pid will be written out by
`rocknix-bluetooth-agent` on start.

I have also cleaned up the call to pgrep in `rocknix-bluetooth` to
ensure that the PID is being set only by the agent.

Tested on a Retroid Pocket 5 by writing bind mounting the two modified
files and starting / stopping bluetooth.

Qapf commented:
Latest now has the agent itself writing its PID. 

@spycat88 is 1000% correct that you need to be patient and count to 5 before going to list devices. If you try to visit that page before the signal handlers get registered, the service is killed. 

I'm open to idea on how to fix the speedrun problem. 

Is there a way to have a UI element "spin" until something happens to declare it done?
Do we move the signal handler registration out of `do_main_loop` and into `__main__` so the signals get registered before the `sleep(5)` that's just ahead of `do_main_loop`?

spycat88 commented:
This issue should now be resolved as of: https://github.com/ROCKNIX/distribution/pull/1208
